compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPULowerCoalescedDMAToGatherLDS.cpp

```diff
@@ -0,0 +1,240 @@
+    SmallVector<int64_t> tileSizes(sourceShape.size(), 1);
+    tileSizes.back() = *subgroupSize * elementsPerTransfer;
+
+    for (SmallVector<int64_t> offsets :
+         StaticTileOffsetRange(sourceShape, tileSizes)) {
+      SmallVector<Value> srcIndices;
+      SmallVector<Value> dstIndices;
+
+      for (auto [dim, offset] : llvm::enumerate(offsets)) {
+        Value offsetVal = arith::ConstantIndexOp::create(rewriter, loc, offset);
+
+        // Build source index for this dimension.
+        Value srcIdx;
+        if (dim < numIndexDims) {
+          // This dimension has an index memref - load from it.
+          srcIdx = memref::LoadOp::create(rewriter, loc, indices[dim],
+                                          ValueRange{offsetVal});
+        } else {
+          // No index memref for this dimension - use the offset directly.
+          srcIdx = offsetVal;
+        }
+
+        // For the innermost dimension, add lane offset to source index.
+        if (dim == sourceShape.size() - 1) {
+          srcIdx = arith::AddIOp::create(rewriter, loc, srcIdx, laneOffset);
+        }
+        srcIndices.push_back(srcIdx);
+
+        // Build destination index (always use the offset directly).
+        dstIndices.push_back(offsetVal);
+      }
+
+      amdgpu::GatherToLDSOp::create(rewriter, loc, source, srcIndices, dest,
+                                    dstIndices, TypeAttr::get(transferType));
+    }
+
+    rewriter.eraseOp(dmaOp);
+    return success();
+  }
+
+private:
+  ArrayRef<int64_t> targetDmaSizes;
+};
+
+namespace {
+struct AMDGPULowerCoalescedDMAToGatherLDSPass final
+    : impl::AMDGPULowerCoalescedDMAToGatherLDSPassBase<
+          AMDGPULowerCoalescedDMAToGatherLDSPass> {
+  void runOnOperation() override {
+    mlir::FunctionOpInterface funcOp = getOperation();
+
+    IREE::GPU::TargetAttr target = getGPUTargetAttr(funcOp);
+    LDBG() << "GPU Target attribute: " << target;
+    if (!target) {
+      LDBG() << "Missing GPU target attribute, pass will fail";
+      // Don't fail if no target attribute - just skip the pass.
+      return;
+    }
+
+    // dma_sizes is optional - if not specified, skip the size validation.
+    ArrayRef<int64_t> dmaSizes;
+    if (auto dmaSizesAttr = target.getWgp().getDmaSizes()) {
+      dmaSizes = dmaSizesAttr.asArrayRef();
+    }
+
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(context);
+    patterns.add<LowerCoalescedGatherDMAPattern>(context, dmaSizes);
+
+    walkAndApplyPatterns(funcOp, std::move(patterns));
+
+    // Verify all CoalescedGatherDMAOps were lowered. This is the current
+    // behavior until we have a fallback lowering path.
+    WalkResult result = funcOp.walk([&](IREE::GPU::CoalescedGatherDMAOp op) {
+      op.emitOpError("failed to lower coalesced_gather_dma op");
+      return WalkResult::interrupt();
+    });
+    if (result.wasInterrupted()) {
+      return signalPassFailure();
+    }
```

Comment:
This should be probably put behind `NDEBUG` only -- we generally don't check that all ops are gone outside of dialect conversion